whereby/dev-docs
# Expose device toggle methods on <whereby-embed />

When https://github.com/whereby/pwa/pull/3015 lands, we will support sending commands to a Whereby window. First two supported commands are `toggle_camera` and `toggle_microphone`.

This change introduces first-class methods of the custom `<whereby-embed />` element to perform these actions, and the methods take care of message passing to the iframe.

## How to test ##
Sadly there is no easy way to test, as I don't want to make too many changes in this repo before moving the custom element to its own location where testing should be made easier. Meanwhile, the following procedure should make you able to test the new functionality:

1. Make sure you have a local embedded organization set up in local-stack which has `https://localhost:8080` in the "Allowed domains" and a transient room with `<room_url>`
2. Start local-stack with PWA PR https://github.com/whereby/pwa/pull/3015 checked out, unless it is merged
3. In this repo, remove the `<meta ..refresh />` in index.html, and replace `import "https://cdn.srv.whereby.com/embed/v1.js";` in tester.js with the local path to whereby-embed.js.
4. Do `yarn start`
5. Open `https://localhost:8080/embed/tester/?room=<room_url>`, stop at review step in precall with device buttons
6. In DevTools, do `const room = document.getElementsByTagName("whereby-embed")[0]`, and `room.toggleCamera(false)`
7. Verify camera button becomes red in precall
8. Try out `room.toggleMicrophone(true|false)` as well, both in precall and once joined in the room
